tracer/src/Datadog.Trace/Baggage.cs

```diff
@@ -0,0 +1,191 @@
+﻿// <copyright file="Baggage.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+
+#nullable enable
+
+namespace Datadog.Trace;
+
+/// <summary>
+/// Baggage is a collection of name-value pairs that are propagated to downstream services.
+/// </summary>
+internal class Baggage
+{
+    private static readonly AsyncLocal<Baggage> AsyncStorage = new();
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="Baggage"/> class.
+    /// </summary>
+    public Baggage()
+    {
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="Baggage"/> class using the specified items.
+    /// </summary>
+    /// <param name="items">The baggage items.</param>
+    public Baggage(IEnumerable<KeyValuePair<string, string>> items)
+    {
+        Items = new ConcurrentDictionary<string, string>(items);
+    }
+
+    private ConcurrentDictionary<string, string>? Items { get; set; }
+
+    /// <summary>
+    /// Gets the baggage collection for the current execution context.
+    /// </summary>
+    public static Baggage Current
+    {
+        get => AsyncStorage.Value ??= new Baggage();
+    }
+
+    /// <summary>
+    /// Gets a value indicating whether this baggage collection is empty.
+    /// </summary>
+    public bool IsEmpty => Items is null or { IsEmpty: true };
+
+    /// <summary>
+    /// Merge the baggage items from <paramref name="b1"/> and
+    /// <paramref name="b2"/> into a new baggage instance.
+    /// </summary>
+    /// <returns>
+    /// A new <see cref="Baggage"/> instance containing all the items
+    /// from <paramref name="b1"/> and <paramref name="b2"/>.
+    /// </returns>
+    /// <remarks>
+    /// If a baggage item with the same key exists in both <paramref name="b1"/> and <paramref name="b2"/>,
+    /// the value from <paramref name="b2"/> will be used.
+    /// </remarks>
+    public static Baggage Merge(Baggage? b1, Baggage? b2)
+    {
+        var items1 = b1?.Items;
+        var items2 = b2?.Items;
+
+        var items1IsEmpty = items1 is null || items1.IsEmpty;
+        var items2IsEmpty = items2 is null || items2.IsEmpty;
+
+        if (items1IsEmpty && items2IsEmpty)
+        {
+            return new Baggage();
+        }
+
+        if (items1IsEmpty)
+        {
+            return new Baggage(items2!);
+        }
+
+        if (items2IsEmpty)
+        {
+            return new Baggage(items1!);
+        }
+
+        var result = new Baggage(items1!);
+        result.Merge(b2);
```

Comment:
Is overriding the value just a side-effect or is that intentional?

Maybe a better question is why we are overriding `b1` with `b2` values (when the same key exists)?

Reply:
If the same key exists in both sets, we need to choose which one has higher priority, or not allow this action. I chose to have `b2` overwrite `b1` since it is the most "recent" baggage, and we want to keep whatever the more recent values are.

This is similar to someone doing:
```csharp
var b = new Baggage();
b.Set("key1", "foo");
b.Set("key1", "bar"); <-- overwrites the old value with the newer one
```

Reply:
fwiw, I realize I don't need this static `Merge()` so I removed it. I kept the instance `Merge` where this is more clear that you are adding `b2` "on top of" what's already there in `b1`
```csharp
b1.Merge(b2);
```